Updating "Component Builds" documentation

MirkoBonadei opened the pull request:
Component builds are not supported by the WebRTC build.
This pull request fixes the documentation with a notice in all the
three pages under 'Native Code'.

Not supporting component builds means that there is no difference
between is_component_build=true and is_component_build=false.

To support this Gn argument we should update all the BUILD.gn files to use
the target type 'component' instead of 'rtc_static_library'/'rtc_shared_library'.
This would give Gn the power to generate shared libraries or static libraries
looking at the value of 'is_component_build'.

This is a non trivial effort and we don't think that the generation
of single WebRTC shared "sub-libraries" is common enough to justify the effort.

These targets are still available on their own but their type is statically
declared in the BUILD.gn file.

MirkoBonadei committed:
Updating "Component Builds" documentation

Component builds are not supported by the WebRTC build.
This pull request fixes the documentation with a notice in all the
three pages under 'Native Code'.

Not supporting component builds means that there is no difference
between is_component_build=true and is_component_build=false.

To support this Gn argument we should update all the BUILD.gn files to use
the target type 'component' instead of 'rtc_static_library'/'rtc_shared_library'.
This would give Gn the power to generate shared libraries or static libraries
looking at the value of 'is_component_build'.

This is a non trivial effort and we don't think that the generation
of single WebRTC shared "sub-libraries" is common enough to justify the effort.

These targets are still available on their own but their type is statically
declared in the BUILD.gn file.

KaptenJansson reviewed (approved):
LGTM

KaptenJansson commented:
Maybe send out a PSA to discuss just to highlight this info since there have been questions about this in the past.

kjellander-unity commented on native-code/android/index.md:
```diff
@@ -49,6 +49,16 @@ the code, building etc.
+     **NOTICE: Component Builds are not supported**
+
+     The Gn arg `is_component_build` is currently ignored for WebRTC builds.
+     [Component builds][5] are supported by Chromium and the arg
```
arg -> argument

kjellander-unity commented on native-code/android/index.md:
```diff
@@ -49,6 +49,16 @@ the code, building etc.
       * To build for 32-bit x86: use `target_cpu="x86"`
       * To build for 64-bit x64: use `target_cpu="x64"`
 
+     **NOTICE: Component Builds are not supported**
```
I'm wondering if it's not better to just leave this out entirely and send a PSA: to discuss-webrtc to just clarify that it isn't and has never been supported. This is quite a lot of text added for something we just don't support.

MirkoBonadei replied:
Yes, I was thinking the same but I don't know if the PSA is enough to avoid questions about this argument in (let's say) one year.

kjellander-unity replied:
Yeah, you're probably right. Let's wait for feedback from gn-dev@ before we decide on this.

kjellander-unity commented on native-code/ios/index.md:
```diff
@@ -85,12 +81,22 @@ with the new arguments.
 
 ~~~~~ bash
 # debug build for 64-bit iOS
-gn gen out/ios_64 --args='target_os="ios" target_cpu="arm64" is_component_build=false'
```
This is great, since it will always be false on iOS anyway: https://cs.chromium.org/chromium/src/build/config/BUILDCONFIG.gn?rcl=40fdca287f3492e18cd4e9ff513c3b846a1d8089&l=164

kjellander-unity reviewed (commented):
Can you investigate if there's a way we can always set it to false for WebRTC? Maybe in our templates?
It defaults to true for Debug builds in Chromium, so it would be better if we set it to false even if it doesn't have any effect.

MirkoBonadei committed:
arg -> argument

MirkoBonadei committed:
Removing component build notice

We have moved this message to Gn. If you try to generate a project with
is_component_build=true then you will get a meaningful error message.

This means that we can remove this notice and make the process easier.

MirkoBonadei committed:
Removing useless links

MirkoBonadei commented:
PTAL. :)

We decided to remove everything about `is_component_build` because now it is defaulted to `false` by WebRTC and there is a check that raises an error if it is set to `true` (with a meaningful error message).

MirkoBonadei commented:
We can probably avoid the PSA because we have never supported this feature and now it is explicitly managed in WebRTC Gn files.

kjellander-unity commented:
This now LGTM. It'll be a good update altogether with https://codereview.webrtc.org/2728643003/
Check if there's any discuss-webrtc thread we can also reply in.
